Self-supervised contrastive learning with SimSiam

sayakpaul opened the pull request:
This example is a minimal walkthrough of how self-supervised contrastive systems work in general in computer vision. For this example, I went with [SimSiam](https://arxiv.org/abs/2011.10566) since it beautifully unifies three other methods from this field - [SimCLR](https://arxiv.org/abs/2002.05709), [BYOL](https://arxiv.org/abs/2006.07733), and [MoCov2](https://arxiv.org/abs/2003.04297). 

I hope the community will benefit from this example and it will make it easier for folks interested to get started in this domain.

sayakpaul commented:
@fchollet just wanted to check if you had the time to review it. Totally understandable if not.

fchollet commented:
Thanks for the PR! I'll review it soon, likely right after we're done with #410.

sayakpaul commented:
Totally understandable. I am sorry, I should have checked before.

fchollet commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+test_ds = test_ds.batch(BATCH_SIZE).prefetch(AUTO)
+
+# Extract the backbone ResNet20.
+backbone = tf.keras.Model(simsiam.encoder.input, simsiam.encoder.layers[-6].output)
```
Consider retrieving the layer by name rather than by index, to make the action more explicit and easier to change

fchollet commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+## Defining the encoder and the predictor
+
+We use an implementation of ResNet20 that is specifically configured for the CIFAR10
+dataset. The code is taken from the
```
What are the benefits of using this model here rather than e.g. `applications.ResNet50`?

I see that the model uses a 32x32 input -- usually that's an antipattern for small image processing. It usually works better to upscale the images first.

https://github.com/GoogleCloudPlatform/keras-idiomatic-programmer/blob/master/zoo/resnet/resnet_cifar10_v2.py#L159

sayakpaul replied:
I had tried with ResNet50. Because of its capacity, it easily leads to representation collapse. This is also studied in all the works I have mentioned throughout this example.

fchollet reviewed (commented):
This is a great topic for an example, and it will make a very valuable addition. The code looks great (just a few minor conerns/comments). Thanks for the PR!

sayakpaul commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+
+# This function is referred from here:
+# https://github.com/google-research/simclr/blob/master/data_util.py.
+def random_apply(func, x, p):
```
It does. Thanks for the suggestion!

sayakpaul commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+    x = tf.image.random_brightness(x, max_delta=0.8 * s[0])
+    x = tf.image.random_contrast(x, lower=1 - 0.8 * s[1], upper=1 + 0.8 * s[1])
+    x = tf.image.random_saturation(x, lower=1 - 0.8 * s[2], upper=1 + 0.8 * s[2])
+    x = tf.image.random_hue(x, max_delta=0.2 * s[3])
```
Notable thoughts. Thank you! 

But as studied in SimCLR and all the works that followed it affine transforms and crops are collectively responsible to prevent the representation collapse situation.

sayakpaul commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+
+def flip_random_crop(image):
+    # With random crops we also apply horizontal flipping.
+    image = tf.image.random_flip_left_right(image)
```
I tried closely following the augmentation pipeline that is typical in the self-supervision literature, hence.

sayakpaul commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+ssl_ds_one = tf.data.Dataset.from_tensor_slices(x_train)
+ssl_ds_one = (
+    ssl_ds_one.shuffle(1024, seed=SEED)
+    .map(custom_augment, num_parallel_calls=AUTO)
```
I had thought about that. But for the following reason, did not go for it:
* When using this code with larger dimensions (e.g. 224x224) and with a bigger dataset, allocating the GPU/TPU solely for model training might be more beneficial.

matpalm replied:
you can also just load the dataset once and just do the augment twice

```
def custom_augment(image):
    # As discussed in the SimCLR paper, the series of augmentation
    # transformations (except for random crops) need to be applied
    # randomly to impose translational invariance.
    image = flip_random_crop(image)
    image = random_apply(color_jitter, image, p=0.8)
    image = random_apply(color_drop, image, p=0.2)
    return image

def augment_twice(image):
    return custom_augment(image), custom_augment(image)

ssl_ds = (tf.data.Dataset.from_tensor_slices(x_train)
    .shuffle(1024, seed=SEED)
    .map(augment_twice, num_parallel_calls=AUTO)
    .batch(BATCH_SIZE)
    .prefetch(AUTO)
)
```

8bitmp3 commented:
I'll also try to take a look asap @sayakpaul Thanks for the awesome example 👍

fchollet commented:
Sorry for the delay, I'll get back to reviewing this soon. I've been pretty swamped.

sayakpaul commented:
No problem and thanks for letting me know.

fchollet commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,435 @@
+representation learning. SSL systems try to formulate a supervised signal from a corpus
+of unlabeled data points.  An example is we train a deep neural network to predict the
+next word from a given set of words. In literature, these tasks are known as *pretext
+tasks* or *auxiliary tasks*. If we [train such a
```
Please make sure to keep all markdown links on a single line to avoid rendering issues

fchollet commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,435 @@
+
+n = 18
+depth = n * 9 + 2
+N_BLOCKS = ((depth - 2) // 9) - 1
```
Style nit: if you go with uppercase for all constnats, let's go with uppercase for `n` and `depth` as well. Also prefer using `num_blocks` to be consistent with naming conventions in the Keras codebase in general.

fchollet reviewed (approved):
Sorry for the delay in reviewing! The code looks great, I just have minor style nits. Thanks!

sayakpaul commented:
No problem, @fchollet. I can only appreciate you taking the time out to review our PRs with such details. 

I have addressed your comments. Let me know if I have the green light on adding the rest of the files.

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+"""
+
+"""
+## Evaluating our SSL method
```
```suggestion
## Evaluating the SSL method
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+the backbone models are usually pre-trained with *pure supervised learning*. 
+"""
+
+# We first create labeled `Dataset` objects.
```
```suggestion
# Create labeled `Dataset` objects.
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+test_ds = tf.data.Dataset.from_tensor_slices((x_test, y_test))
+
+# Then we shuffle, batch, and prefetch this dataset for performance. We
```
```suggestion
# Shuffle, batch, and prefetch this dataset for performance;
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+)
+test_ds = test_ds.batch(BATCH_SIZE).prefetch(AUTO)
+
+# Extract the backbone ResNet20.
```
```suggestion
# Extract the backbone ResNet-20.
```

sayakpaul replied:
ResNet50 is quite commonly used and so are any other variants of ResNet.

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+    simsiam.encoder.input, simsiam.encoder.get_layer("backbone_pool").output
+)
+
+# We then create our linear classifier and train it.
```
```suggestion
# Create the linear classifier and instantiate the model.
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+outputs = layers.Dense(10, activation="softmax")(x)
+linear_model = tf.keras.Model(inputs, outputs, name="linear_model")
+
+# Start training.
```
```suggestion
# Compile the model and begin training it.
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+[fine-tuning](https://keras.io/guides/transfer_learning/) on the source dataset or even a
+target dataset with 5% or 10% labels present. Practically, we can use the backbone model
+for any downstream task such as semantic segmentation, object detection, and so on where
+the backbone models are usually pre-trained with *pure supervised learning*. 
```
Suggestion: some refactoring

```suggestion
The most popular method for evaluating an SSL method in computer vision (or any
other pre-training method as such) is to learn a linear classifier on the frozen
features of the trained backbone model (in our case, it's ResNet-20), and then
evaluate the classifier on unseen images. Other methods include
[fine-tuning](https://keras.io/guides/transfer_learning/) on the source dataset
or even a target dataset with 5% or 10% labels present. We can use the backbone
model for any downstream task, such as semantic segmentation, object detection,
and so on, where the backbone models are usually pre-trained with *pure
supervised learning*. 
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+   * [Self-supervised learning: The dark matter of
+intelligence](https://ai.facebook.com/blog/self-supervised-learning-the-dark-matter-of-intelligence/)
+   * [Understanding self-supervised learning using controlled datasets with known
+structure](https://sslneuips20.github.io/files/CameraReadys%203-77/64/CameraReady/Understanding_self_supervised_learning.pdf)
```
Suggestion: some refactoring

```suggestion
## Notes

* More data and longer pre-training schedule can benefit SSL in general.
* SSL is particularly helpful when you have very limited *labeled* training
data but you can have a large corpus of unlabeled data. In an SSL method called
[SwAV](https://arxiv.org/abs/2006.09882), researchers trained a
[RegNet](https://arxiv.org/abs/2006.09882) on 2 billion images—their results in
downstream performance was very close to the one achieved by pure supervised
pre-training. In addition, for certain downstream tasks, their method managed to
outperform the supervised counterparts. (For more details, check 
[Goyal et al., 2021](https://arxiv.org/pdf/2103.01988.pdf).

To learn more about contrastive SSL and how it can help neural networks learn
meaningful representations:

* [Self-supervised learning: The dark matter of
intelligence](https://ai.facebook.com/blog/self-supervised-learning-the-dark-matter-of-intelligence/).
* [Understanding self-supervised learning using controlled datasets with known
structure](https://sslneuips20.github.io/files/CameraReadys%203-77/64/CameraReady/Understanding_self_supervised_learning.pdf).
```

@sayakpaul LMK if I got the gist right here!

8bitmp3 reviewed (commented):
This is an awesome example @sayakpaul PTAL and LMKWYT

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,431 @@
+Description: Implementation of a self-supervised learning method for computer vision.
+"""
+"""
+Self-supervised learning (SSL) is an interesting branch of study in the field of
+representation learning. SSL systems try to formulate a supervised signal from a corpus
+of unlabeled data points.  An example is we train a deep neural network to predict the
+next word from a given set of words. In literature, these tasks are known as *pretext
+tasks* or *auxiliary tasks*. If we [train such a network](https://arxiv.org/abs/1801.06146) on a huge dataset (such as
+the [Wikipedia text corpus](https://www.corpusdata.org/wikipedia.asp)) it learns very effective
+representations that transfer well to downstream tasks. Language models like
+[BERT](https://arxiv.org/abs/1810.04805), [GPT-3](https://arxiv.org/abs/2005.14165),
+[ELMo](https://allennlp.org/elmo) all benefit from this.
```
Suggestion:

- Describe the example in the first few sentences (and then later describe the steps, as you have already) (SimSiam on CIFAR-10).
- Shorten a few sentences (be mindful of the users attention span) without affecting the quality
- Some refactoring
- Link ELMo to https://arxiv.org/abs/1802.05365

For example:

```suggestion
Self-supervised learning (SSL)—a branch of study in representation learning—
tries to formulate a supervised signal from a corpus of unlabeled data points.
This example demonstrates how to implement one such SSL system—
simple Siamese networks (SimSiam) 
([Chen & He, 2020](https://arxiv.org/abs/2011.10566))—on the
[CIFAR-10](https://www.cs.toronto.edu/~kriz/cifar.html) dataset.
SSL can be applied to natural language modeling tasks, such as predicting the
next word from a given set of words, which is also known as *pretext tasks* or
*auxiliary tasks*. If we train a neural network, such as ULMFiT 
([Howard & Ruder, 2018](https://arxiv.org/abs/1801.06146)),
on the [Wikipedia text corpus](https://www.corpusdata.org/wikipedia.asp)),
the model can learn effective representations that can transfer well to
downstream tasks. Notable beneficiaries of this methodology are language models
like BERT ([Devlin et al., 2018](https://arxiv.org/abs/1810.04805)), GPT-3
([Brown et al., 2020](https://arxiv.org/abs/2005.14165)), and ELMo
([Peters et al., 2018](https://arxiv.org/abs/1802.05365)).
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,431 @@
+[BERT](https://arxiv.org/abs/1810.04805), [GPT-3](https://arxiv.org/abs/2005.14165),
+[ELMo](https://allennlp.org/elmo) all benefit from this.
+
+Much like the language models we can train computer vision models using similar
+approaches. To make things work in computer vision, we need to formulate the learning
+tasks such that the underlying model (a deep neural network) is able to make sense of the
+semantic information present in vision data. One such task is to a model to _contrast_
+between two different versions of the same image. The hope is that in this way the model
+will have learn representations where the similar images are grouped as together possible
+while the dissimilar images are further away.
```
Suggestion:

- Refactor (repetition of "computer vision")
- Shorten a few sentences (be mindful of the users attention span) without affecting the quality

```suggestion
We can also train computer vision models using similar SSL approaches. The model
(a deep neural network) needs to make sense of the semantic information present
in vision data, and one way of doing that is modeling to _contrast_ between two
different versions of the same image. This way, the model can attempt to learn
representations where similar images are grouped as "together" as possible,
while dissimilar images aren't.
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,431 @@
+"""
+## Defining our data augmentation pipeline
+
+As studied in [SimCLR](https://arxiv.org/abs/2002.05709) having the right data
+augmentation pipeline is critical for SSL systems to work effectively in computer vision.
+Two particular augmentation transforms that seem to matter the most are: 1.) Random
+resized crops and 2.) Color distortions. Most of the other SSL systems for computer
+vision (such as [BYOL](https://arxiv.org/abs/2006.07733),
+[MoCoV2](https://arxiv.org/abs/2003.04297), [SwAV](https://arxiv.org/abs/2006.09882),
+etc.) include these in their training pipelines.
```
Suggestion:

- Some refactoring
- Add author references

```suggestion
Having the right data augmentation pipeline can be critical for the SSL systems
to work effectively in computer vision tasks, as demonstrated in SimCLR
([Chen et al., 2020](https://arxiv.org/abs/2002.05709)). Two notable augmentation transforms are: 

1. Random resized crops; and 
2. Color distortions. 

Other SSL systems for computer vision, such as BYOL
([Grill et al., 2020](https://arxiv.org/abs/2006.07733)), MoCoV2
([Chen et al., 2020](https://arxiv.org/abs/2003.04297)), and SwAV 
[(Caron et al., 2020)](https://arxiv.org/abs/2006.09882)),
make use of these augmentation techniques in their training pipelines.
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,431 @@
+"""
+## Defining the encoder and the predictor
+
+We use an implementation of ResNet20 that is specifically configured for the CIFAR10
```
```suggestion
We use an implementation of ResNet-20 that is specifically configured for the CIFAR-10
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,431 @@
+"""
+## Evaluating our SSL method
+
+The most popularly used method to evaluate a SSL method in computer vision (or any other
+pre-training method as such) is to learn a linear classifier on the frozen features of
+the trained backbone model (in this case it is ResNet20) and evaluate the classifier on
+unseen images. Other methods include
+[fine-tuning](https://keras.io/guides/transfer_learning/) on the source dataset or even a
+target dataset with 5% or 10% labels present. Practically, we can use the backbone model
+for any downstream task such as semantic segmentation, object detection, and so on where
+the backbone models are usually pre-trained with *pure supervised learning*.
```
Suggestion:

```suggestion
The most popular method for evaluating an SSL method in computer vision (or any
other pre-training method as such) is to learn a linear classifier on the frozen
features of the trained backbone model (in our case, it's ResNet-20), and then
evaluate the classifier on unseen images. Other methods include
[fine-tuning](https://keras.io/guides/transfer_learning/) on the source dataset
or even a target dataset with 5% or 10% labels present. We can use the backbone
model for any downstream task, such as semantic segmentation, object detection,
and so on, where the backbone models are usually pre-trained with *pure
supervised learning*. 
```

8bitmp3 commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,431 @@
+print("Test accuracy: {:.2f}%".format(test_acc * 100))
+
+"""
+## Notes
+* More data and longer pre-training schedule benefit SSL in general.
+* SSL is particularly very helpful when you do not have access to very limited *labeled*
+training data but you can manage to build a large corpus of unlabeled data. Recently,
+using an SSL method called [SwAV](https://arxiv.org/abs/2006.09882), a group of
+researchers at Facebook trained a [RegNet](https://arxiv.org/abs/2006.09882) on 2 Billion
+images. They were able to achieve downstream performance very close to those achieved by
+pure supervised pre-training. For some downstream tasks, their method even outperformed
+the supervised counterparts. You can check out [their
+paper](https://arxiv.org/pdf/2103.01988.pdf) to know the details.
+* If you are interested to understand why contrastive SSL helps networks learn meaningful
+representations, you can check out the following resources:
+   * [Self-supervised learning: The dark matter of
+intelligence](https://ai.facebook.com/blog/self-supervised-learning-the-dark-matter-of-intelligence/)
+   * [Understanding self-supervised learning using controlled datasets with known
+structure](https://sslneuips20.github.io/files/CameraReadys%203-77/64/CameraReady/Understanding_self_supervised_learning.pdf)
```
Suggestion:

```suggestion
## Notes

* More data and longer pre-training schedule can benefit SSL in general.
* SSL is particularly helpful when you have very limited *labeled* training
data but you can have a large corpus of unlabeled data. In an SSL method called
[SwAV](https://arxiv.org/abs/2006.09882), researchers trained a
[RegNet](https://arxiv.org/abs/2006.09882) on 2 billion images—their results in
downstream performance was very close to the one achieved by pure supervised
pre-training. In addition, for certain downstream tasks, their method managed to
outperform the supervised counterparts. (For more details, check 
[Goyal et al., 2021](https://arxiv.org/pdf/2103.01988.pdf).

To learn more about contrastive SSL and how it can help neural networks learn
meaningful representations, check out the following posts:

* [Self-supervised learning: The dark matter of
intelligence](https://ai.facebook.com/blog/self-supervised-learning-the-dark-matter-of-intelligence/).
* [Understanding self-supervised learning using controlled datasets with known
structure](https://sslneuips20.github.io/files/CameraReadys%203-77/64/CameraReady/Understanding_self_supervised_learning.pdf).
```

sayakpaul commented on examples/vision/simsiam.py:
```diff
@@ -0,0 +1,434 @@
+the [Wikipedia text corpus](https://www.corpusdata.org/wikipedia.asp)) it learns very effective
+representations that transfer well to downstream tasks. Language models like
+[BERT](https://arxiv.org/abs/1810.04805), [GPT-3](https://arxiv.org/abs/2005.14165),
+[ELMo](https://allennlp.org/elmo) all benefit from this. 
```
For this example, I would go with the initial one I had written. 

> SSL can be applied to natural language modeling tasks,

This directly states SSL can be applied to NLP.  I prefer to build a premise first and let the user flow into that naturally.

fchollet reviewed (approved):
Thanks for the great contribution! 👍

ktjayamanna commented:
I ran the simsiam example here without changing anything on a V100 GPU for 800 epochs. Can't go beyond 36% ish accuracy. Is that the case for others as well?

sayakpaul commented:
Hi. 

The purpose of the example was to walk the readers through a workflow and not reproduce the entire result per se. 

Here's an extended discussion of you want to follow more:

https://discuss.tensorflow.org/t/self-supervised-contrastive-learning-with-simsiam/263/2

preesee commented:
@ktjayamanna I have the same issue with this example. The performance is lower than 30% after 200 epochs. I guess that the backbone of the Resnet is not pre-trained correctly in this model. I had to change to test other examples.

sayakpaul commented:
I myself concur with that. I think I have made the objectives of the tutorial and differences in the implementation abundantly clear.
